contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -1687,3 +1695,113 @@ allocates resources such as `NodePorts` and `ClusterIPs` and automatically fill
+* PersistentVolumeClaims: Users may request a specific PersistentVolume in
+  `spec` or will be allocated one (in the same `spec` field).  If a specific PV
+  is requested, the volume controller will either ensure that the volume is
+  available or report failure asynchronously.  Consumers read the result by
+  examining both the PVC and the PV.  This is more complicated than the others
+  because the `spec` value is stored before being confirmed, which could lead
+  to a user accessing someone else's PV.
+
+A counter-example:
+
+* Service `externalIPs`: Users must specify one or more specific IPs in `spec`.
+  The system cannot easily verify those IPs (by their definition, they are
+  external). Consumers read the result from `spec`.  This is UNSAFE and has
+  caused problems with untrusted users.
+
+In the past, API conventions dictated that `status` fields always come from
+observation, which made some of these cases more complicated than necessary.
+The conventions have been updated to allow `status` to hold such allocated
+values.  This is not a one-size-fits-all solution, though.
+
+### When to use a `spec` field
+
+New APIs should almost never do this.  Instead, they should use `status`.
+PersistentVolumes might have been simpler if we had done this.
+
+### When to use a `status` field
+
+Storing such values in `status` is the easiest and most straight-forward
+pattern.  This is appropriate when:
+
+* the allocated value is highly coupled to the rest of the object (e.g. pod
+  resource allocations)
+* the allocated value is always or almost always needed (i.e. most instances of
+  this type will have a value)
+* the schema and controller are known a priori (i.e. it's not an extension)
+* it is "safe" to allow the controller(s) to write to `status` (i.e.
+  there's low risk of them causing problems via other `status` fields).
+
+Consumers of such values can look at the `status` field for the "final" value.
+
+#### Sequencing operations
+
+Since almost everything is happening asynchronously to almost everything else,
+controller implementations should take care around the ordering of operations.
+For example, whether the controller updates a `status` field before or after it
+actuates a change depends on what guarantees need to be made to observers of
+the system.  In some cases, writing to a `status` field represents an
+acknowledgement or acceptance of a `spec` value, and it is OK to write it before
+actuation.  However, if it would be problematic for a client to observe the
+`status` value before it is actuated then the controller must actuate first and
+update `status` afterward.  In some rarer cases, controllers will need to
+acknowledge, then actuate, then update to a "final" value.
+
+Controllers must take care to consider how a `status` field will be handled in
+the case of interrupted control loops (e.g. controller crash and restart), and
+must act idempotently and consistently.
```

Comment:
Consistency is interesting when the main way controllers observe data and react is usually via an informer-fed cache. The typical pattern of reading from spec and writing to status (happens to) avoid stomping status if your observed data is stale.

Starting to both write to and read from status as an input for making decisions can lead controllers into bugs where their informer-fed cache has not yet observed their own earlier write, and they make a bad/weird decision and persist the result to status.

An example we hit recently where a controller was using status both as a place to record information and as an input for decisions was a [bug in the job controller](https://github.com/kubernetes/kubernetes/issues/105199) where it was writing completed pod uids to job status, then reading completed pod uids from job status to aggregate them into an overall success/failure count after having marked the pods as observed (to try to be resilient/idempotent).

The best mitigation for a controller is usually to use a resourceVersion precondition and react to a conflict response that indicates the view of the object you had was stale. Controllers that expect to be the only one acting on their object's status may not be accustomed to using that approach. I'd recommend highlighting (or linking to a discussion of) using resourceVersion preconditions for status update or patch requests, *especially* when using status as both an output and an input.

Reply:
ACK.  Will link the above issue.